### ISSUE TITLE
Linking videoio fails on macOS

### ISSUE THREAD

**Issue by basvandijk** [opencv/opencv#13283]
##### System information (version)

- OpenCV => 4.0.0
- Operating System / Platform => nixpkgs on macOS 10.14 (Mojave)
- Compiler => clang

##### Detailed description

I'm packaging OpenCV-4.0.0 in nixpkgs at https://github.com/NixOS/nixpkgs/pull/51027.

The build succeeds on the `aarch64-linux` and `x86_64-linux` platforms. Unfortunately it fails on `x86_64-darwin` while linking the `videoio` module. See the full build log [here](https://logs.nix.ci/?key=nixos/nixpkgs.51027&attempt_id=2f1211fc-1c0b-45fb-837f-f240cc720170).

So it looks like a linker issue:

```
[ 62%] Linking CXX shared library ../../lib/libopencv_videoio.dylib
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_NSData", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
  "_OBJC_CLASS_$_NSDate", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
  "_OBJC_CLASS_$_NSDictionary", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
  "_OBJC_CLASS_$_NSURL", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
ld: symbol(s) not found for architecture x86_64
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [modules/videoio/CMakeFiles/opencv_videoio.dir/build.make:193: lib/libopencv_videoio.4.0.0.dylib] Error 1
```

Apparently [`NSData` is part of the Foundation framework](https://developer.apple.com/documentation/foundation/nsdata). I don't see `-framework foundation` anywhere in the opencv sources so I tried adding it to the makefile of the failing videoio module [here](https://github.com/opencv/opencv/blob/4.0.0/modules/videoio/CMakeLists.txt#L16). Unfortunately that didn't help.

Any idea what I'm doing wrong?

##### Steps to reproduce

First [install Nix](https://nixos.org/nix/) on OS X.

```
git clone git@github.com:NixOS/nixpkgs.git
cd nixpkgs

# Fetch my PR that adds OpenCV-4.0.0
git fetch origin pull/51027/head:opencv-4.0.0
git checkout opencv-4.0.0

# Try building OpenCV-4.0.0
nix-build -A opencv4
```

**Comment by mshabunin** [comment]
AVFoundation is being linked here: https://github.com/opencv/opencv/blob/f07856eab936dcae9a490acc4ba8ced0db91ca9e/modules/videoio/CMakeLists.txt#L188-L189

Check warnings in your build log:
```
In file included from /tmp/nix-build-opencv-4.0.0.drv-0/source/modules/videoio/src/cap_avfoundation_mac.mm:46:
In file included from /nix/store/7gdrh1bdhnmpibx13c33mhsnm9dl3bnf-apple-framework-AVFoundation/Library/Frameworks/AVFoundation.framework/Headers/AVFoundation.h:12:
In file included from /nix/store/7gdrh1bdhnmpibx13c33mhsnm9dl3bnf-apple-framework-AVFoundation/Library/Frameworks/AVFoundation.framework/Headers/AVBase.h:11:
In file included from /nix/store/fwvbsvbr42dh95i722k9323cl0hkklk2-apple-framework-Foundation/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:247:
/nix/store/lc6ri0hwwszaisws9ifqa5xqflshs3dr-swift-corefoundation/Library/Frameworks/CoreFoundation.framework/Headers/CFAvailability.h:164:5: warning: 'DEPLOYMENT_RUNTIME_SWIFT' is not defined, evaluates to 0 [-Wundef]
#if DEPLOYMENT_RUNTIME_SWIFT
    ^
```
Probably your compiler is not fully compatible with OSX and/or AVFoundation. There are no warnings in [our builds](https://pullrequest.opencv.org/buildbot/builders/precommit_macosx/builds/18460/steps/compile%20release/logs/stdio) with Apple Clang.

**Comment by basvandijk** [comment]
It turned out to be a nixpkgs issue. I fixed it using [this change](https://github.com/NixOS/nixpkgs/compare/e631423a4e9f102590eaa6bc73d2f90caaca7426..7377665ed260252202b6b862f8cd1c868e3dcea2).
